README.md
# Link Validator
 This Confluence add-on checks provided URLs to ensure they are valid - that is, that they don't return an error when accessed.

## link-validator macro
 Checks the provided link to determine if it is a valid location.

 By default, the final status message is available by hovering your mouse over the icon. If you wish to have it displayed on the page directly, specify verbose=true when using the macro.

### Parameters
 | Name | Required | Default Value | Description |
 |------|----------|---------------|-------------|
 | [default]/url | yes | | The URL to validate. This may alternately be the body of the macro, which allows the URL to be generated by other macros. |
 | verbose | no | false | If `true`, the status message will be displayed as text on the page. Otherwise, it is available by hovering the mouse over the status icon. | 
 | timeout | no | 5 | The `timeout` for each request. If a request fails it is retried two additional times to make sure, which means that the actual wait time for a non-responsive URL is `timeout` X 3.  |
 
### Wiki markup
 ```
 {link-validator:url=http://address}{link-validator}
 {link-validator}http://address{link-validator}
```


CHANGES.md
Changes
=======
2017-03-27, 4.3.0, 5.8.2-6.1.0
------------------------------
* LINKVAL-74 Confluence 6.1.0 compatibility.

2017-02-28, 4.2.0, 5.8.2-6.0.6
------------------------------
* LINKVAL-66 Confluence 6.0.6 compatibility.

2016-07-05, 4.1.0, 5.5-5.10.1
-----------------------------
* AN56-63 Confluence 5.10.1 compatibility.

2016-05-23, 4.0.0, 5.5-5.9.11
-----------------------------
* AN56-54 Remove RandomBits dependencies.

2015-12-08, 3.0.0, 5.4-5.9
----------------------------
* AN56-45 Update RB to compile time library and Confluence 5.9.2 compability.

2014-09-24, 2.1.3, 5.3-5.6
--------------------------
* AN56-35 - Compatible with Confluence 5.6

2014-05-23, 2.1.2, 5.3-5.5.1
----------------------------
* Confluence 5.5 compatibility

2013-10-11, 2.1.1, 5.2
----------------------
* Fixed installation problem in Confluence 5.2
* Added integration test


src/test/java/net/customware/confluence/plugin/linkvalidator/LinkValidatorUtilTest.java
package net.customware.confluence.plugin.linkvalidator;

import com.atlassian.confluence.setup.BootstrapManager;
import com.atlassian.spring.container.ContainerContext;
import com.atlassian.spring.container.ContainerManager;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.MockitoAnnotations.Mock;
import org.mockito.runners.MockitoJUnitRunner;

import static net.customware.confluence.plugin.linkvalidator.LinkValidatorUtil.createGoodIcon;
import static net.customware.confluence.plugin.linkvalidator.LinkValidatorUtil.createInfoIcon;
import static org.hamcrest.Matchers.containsString;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.when;

/**
 * @author yasir
 * @author yclian
 * @since 2.1.0-SNAPSHOT-01042013
 * @version 2.1.0-SNAPSHOT-01042013
 */
@RunWith(MockitoJUnitRunner.class)
public class LinkValidatorUtilTest {

    @Mock BootstrapManager bootstrapManager;
    @Mock ContainerContext containerContext;

    @Before public void setUpContext() {

        ContainerManager.getInstance().setContainerContext(containerContext);

        when(containerContext.getComponent("bootstrapManager")).thenReturn(bootstrapManager);
        when(bootstrapManager.getWebAppContextPath()).thenReturn("");
    }

    @After public void tearDownContext() {
        ContainerManager.getInstance().setContainerContext(null);
    }

    @Test public void createIcons() {

        String goodIcon = createGoodIcon("good", true);
        String infoIcon = createInfoIcon("The information is informative", true);

        assertThat(goodIcon, containsString("<img src='/images/icons/emoticons/check.png' title='good'/> good"));
        assertThat(infoIcon, containsString("<img src='/images/icons/emoticons/information.png' title='The information is informative'/> The information is informative"));
    }
}


src/test/java/it/net/customware/confluence/plugin/linkvalidator/StatelessTestRunner.java
package it.net.customware.confluence.plugin.linkvalidator;

import com.atlassian.confluence.api.model.content.*;
import com.atlassian.confluence.test.api.model.person.UserWithDetails;

import java.util.List;

/**
 * @author khailoon
 * @since 4.3.0.20170329
 */
public interface StatelessTestRunner {

    /**
     * Create page {@link Content} in the default space
     * @param title the title of the page
     * @param content content of the page
     * @param contentRepresentation {@link ContentRepresentation}
     * @return the created page {@link Content}
     */
    Content createPage(String title, String content, ContentRepresentation contentRepresentation);

    /**
     * Create page {@link Content}
     * @param title the title of the page
     * @param content content of the page
     * @param contentRepresentation {@link ContentRepresentation}
     * @param space {@link Space} to create the new content
     * @return the created page {@link Content}
     */
    Content createPage(String title, String content, ContentRepresentation contentRepresentation, Space space);

    Content createPageForUser(UserWithDetails user, String title, String content, ContentRepresentation contentRepresentation);

    Content createPageForUser(UserWithDetails user, String title, String content, ContentRepresentation contentRepresentation, Space space);

    Content anonymousCreatePage(String title, String content, ContentRepresentation contentRepresentation);

    Content editPageContent(Content content, String newContent, ContentRepresentation contentRepresentation);

    Content editPageContent(Content content, String newTitle, String newContent, ContentRepresentation contentRepresentation);

    Content createChildPage(String title, String content, ContentRepresentation contentRepresentation, Content parent);

    List<Label> createLabels(String labels, String separator);

    void applyLabelsToContent(Content content, String labels);

    void removeContent(Content content);

    Content getViewPage(Content page);

    Content getViewPageOfVersion(Content page, Integer version);

    Content getPageByPageTitle(String pageTitle);

    Content getViewPageAsAnonymous(Content page);

    String getRenderedContent(Content viewPage);

    String getHTMLContent(Content viewPage);

    List<Space> createSpaces(Integer noOfSpaces);

    Space createSpace(String key, String name);

    void removeSpaces(List<Space> spaces);

    void removeSpace(Space space);

    void applyLabelToSpace(Space space, String label);

    Integer getSpaceCount();

    List<AttachmentUpload> createAttachments(String filePaths, String separator);

    Content addAttachmentToPage(Content page, String filePaths);

    void waitForIndexingToComplete();

    String convertWikiToStorage(String wiki);

    void visitEditContentPage(Content content);

    void bindViewPage();
}


src/test/java/it/net/customware/confluence/plugin/linkvalidator/LinkValidatorMacroTest.java
package it.net.customware.confluence.plugin.linkvalidator;

import com.atlassian.confluence.api.model.content.Content;
import com.atlassian.confluence.api.model.content.ContentRepresentation;
import com.atlassian.confluence.webdriver.pageobjects.page.content.ViewPage;
import org.junit.Test;

import java.util.concurrent.TimeUnit;

import static org.hamcrest.CoreMatchers.containsString;
import static org.junit.Assert.assertThat;

public class LinkValidatorMacroTest extends DefaultStatelessTestRunner {
    @Test public void shouldSuccessIfValidLink() {
        Content content = createPage("shouldSuccessIfValidLink", "{link-validator:url=" + product.getProductInstance().getBaseUrl() + "|verbose=true}{link-validator}", ContentRepresentation.WIKI);
        ViewPage viewPage = product.viewPage(content);

        webDriver.manage().timeouts().implicitlyWait(10, TimeUnit.SECONDS);
        assertThat(viewPage.getMainContent().getText(), containsString("OK (200)"));
    }

    @Test public void shouldFailIfInvalidLink() {
        Content content = createPage("shouldFailIfInvalidLink", "{link-validator:url=" + product.getProductInstance().getBaseUrl() + "/fail|verbose=true}{link-validator}", ContentRepresentation.WIKI);
        ViewPage viewPage = product.viewPage(content);

        webDriver.manage().timeouts().implicitlyWait(10, TimeUnit.SECONDS);
        assertThat(viewPage.getMainContent().getText(), containsString("Not Found (404)"));
    }
}


src/main/resources/net/customware/confluence/plugin/linkvalidator/js/behaviour.js
/*
 * Copyright (c) 2007, CustomWare Asia Pacific
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *     * Redistributions of source code must retain the above copyright notice,
 *       this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of "CustomWare Asia Pacific" nor the names of its contributors
 *       may be used to endorse or promote products derived from this software
 *       without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

/*
   Behaviour v1.1 by Ben Nolan, June 2005. Based largely on the work
   of Simon Willison (see comments by Simon below).

   Description:
   	
   	Uses css selectors to apply javascript behaviours to enable
   	unobtrusive javascript in html documents.
   	
   Usage:   
   
	var myrules = {
		'b.someclass' : function(element){
			element.onclick = function(){
				alert(this.innerHTML);
			}
		},
		'#someid u' : function(element){
			element.onmouseover = function(){
				this.innerHTML = "BLAH!";
			}
		}
	};
	
	Behaviour.register(myrules);
	
	// Call Behaviour.apply() to re-apply the rules (if you
	// update the dom, etc).

   License:
   
   	My stuff is BSD licensed. Not sure about Simon's.
   	
   More information:
   	
   	http://ripcord.co.nz/behaviour/
   
*/
//////////////// Start of presence check ///////////////
if(typeof Behaviour=='undefined') {

var Behaviour = {
	list : new Array,
	
	register : function(sheet){
		Behaviour.list.push(sheet);
	},
	
	start : function(){
		Behaviour.addLoadEvent(function(){
			// quit if this function has already been called
			if (arguments.callee.done) return;
			arguments.callee.done = true;

			Behaviour.apply();
		});
	},
	
	apply : function(){
	
		var h, i, sheet, selector, list, element;
		for (h=0; sheet=Behaviour.list[h];h++){
			for (selector in sheet){
				list = document.getElementsBySelector(selector);
				
				if (!list){
					continue;
				}

				for (i=0;element=list[i];i++){
					sheet[selector](element);
				}
			}
		}
	},
	
	addLoadEvent : function(func){
		if (window.addEventListener) {
			var load = function() {
				if (arguments.callee.done)
					return;
				arguments.callee.done = true;
				arguments.callee.func();
			}
			load.func = func;
			window.addEventListener("DOMContentLoaded", load, false);
			window.addEventListener("load", load, false);
		} else if (window.attachEvent) {
			window.attachEvent('onload',func);
		} else {
	
			var oldonload = window.onload;
		
			if (typeof window.onload != 'function') {
				window.onload = func;
			} else {
				window.onload = function() {
					oldonload();
					func();
				}
			}
		}
	}
}
if (navigator.appName != "Microsoft Internet Explorer") {
Behaviour.start();
}

/*
   The following code is Copyright (C) Simon Willison 2004.

   document.getElementsBySelector(selector)
   - returns an array of element objects from the current document
     matching the CSS selector. Selectors can contain element names, 
     class names and ids and can be nested. For example:
     
       elements = document.getElementsBySelect('div#main p a.external')
     
     Will return an array of all 'a' elements with 'external' in their 
     class attribute that are contained inside 'p' elements that are 
     contained inside the 'div' element which has id="main"

   New in version 0.4: Support for CSS2 and CSS3 attribute selectors:
   See http://www.w3.org/TR/css3-selectors/#attribute-selectors

   Version 0.4 - Simon Willison, March 25th 2003
   -- Works in Phoenix 0.5, Mozilla 1.3, Opera 7, Internet Explorer 6, Internet Explorer 5 on Windows
   -- Opera 7 fails 
*/

function getAllChildren(e) {
  // Returns all children of element. Workaround required for IE5/Windows. Ugh.
  return e.all ? e.all : e.getElementsByTagName('*');
}

document.getElementsBySelector = function(selector) {
  // Attempt to fail gracefully in lesser browsers
  if (!document.getElementsByTagName) {
    return new Array();
  }
  // Split selector in to tokens
  var token, bits, tagName, id, className, found, foundCount, elements, currentContextIndex, h, j, k;
  var tokens = selector.split(' ');
  var currentContext = new Array(document);
  for (var i = 0; i < tokens.length; i++) {
    token = tokens[i].replace(/^\s+/,'').replace(/\s+$/,'');;
    if (token.indexOf('#') > -1) {
      // Token is an ID selector
      bits = token.split('#');
      tagName = bits[0];
      id = bits[1];
      element = document.getElementById(id);
      if (tagName && element.nodeName.toLowerCase() != tagName) {
        // tag with that ID not found, return false
        return new Array();
      }
      // Set currentContext to contain just this element
      currentContext = new Array(element);
      continue; // Skip to next token
    }
    if (token.indexOf('.') > -1) {
      // Token contains a class selector
      bits = token.split('.');
      tagName = bits[0];
      className = bits[1];
      if (!tagName) {
        tagName = '*';
      }
      // Get elements matching tag, filter them for class selector
      found = new Array;
      foundCount = 0;
      for (h = 0; h < currentContext.length; h++) {
        if (tagName == '*') {
            elements = getAllChildren(currentContext[h]);
        } else {
            elements = currentContext[h].getElementsByTagName(tagName);
        }
        for (j = 0; j < elements.length; j++) {
          found[foundCount++] = elements[j];
        }
      }
      currentContext = new Array;
      currentContextIndex = 0;
      for (k = 0; k < found.length; k++) {
        if (typeof found[k].className === "string" && found[k].className.match(new RegExp('\\b'+className+'\\b'))) {
          currentContext[currentContextIndex++] = found[k];
        }
      }
      continue; // Skip to next token
    }
    // Code to deal with attribute selectors
    if (token.match(/^(\w*)\[(\w+)([=~\|\^\$\*]?)=?"?([^\]"]*)"?\]$/)) {
      tagName = RegExp.$1;
      var attrName = RegExp.$2;
      var attrOperator = RegExp.$3;
      var attrValue = RegExp.$4;
      if (!tagName) {
        tagName = '*';
      }
      // Grab all of the tagName elements within current context
      found = new Array;
      foundCount = 0;
      for (h = 0; h < currentContext.length; h++) {
        if (tagName == '*') {
            elements = getAllChildren(currentContext[h]);
        } else {
            elements = currentContext[h].getElementsByTagName(tagName);
        }
        for (j = 0; j < elements.length; j++) {
          found[foundCount++] = elements[j];
        }
      }
      currentContext = new Array;
      currentContextIndex = 0;
      var checkFunction; // This function will be used to filter the elements
      switch (attrOperator) {
        case '=': // Equality
          checkFunction = function(e) { return (e.getAttribute(attrName) == attrValue); };
          break;
        case '~': // Match one of spaces seperated words
          checkFunction = function(e) { return (e.getAttribute(attrName).match(new RegExp('\\b'+attrValue+'\\b'))); };
          break;
        case '|': // Match start with value followed by optional hyphen
          checkFunction = function(e) { return (e.getAttribute(attrName).match(new RegExp('^'+attrValue+'-?'))); };
          break;
        case '^': // Match starts with value
          checkFunction = function(e) { return (e.getAttribute(attrName).indexOf(attrValue) == 0); };
          break;
        case '$': // Match ends with value - fails with "Warning" in Opera 7
          checkFunction = function(e) { return (e.getAttribute(attrName).lastIndexOf(attrValue) == e.getAttribute(attrName).length - attrValue.length); };
          break;
        case '*': // Match ends with value
          checkFunction = function(e) { return (e.getAttribute(attrName).indexOf(attrValue) > -1); };
          break;
        default :
          // Just test for existence of attribute
          checkFunction = function(e) { return e.getAttribute(attrName); };
      }
      currentContext = new Array;
      currentContextIndex = 0;
      for (k = 0; k < found.length; k++) {
        if (checkFunction(found[k])) {
          currentContext[currentContextIndex++] = found[k];
        }
      }
      // alert('Attribute Selector: '+tagName+' '+attrName+' '+attrOperator+' '+attrValue);
      continue; // Skip to next token
    }
    
    if (!currentContext[0]){
    	return null;
    }
    
    // If we get here, token is JUST an element (not a class or ID selector)
    tagName = token;
    found = new Array;
    foundCount = 0;
    for (h = 0; h < currentContext.length; h++) {
      elements = currentContext[h].getElementsByTagName(tagName);
      for (j = 0; j < elements.length; j++) {
        found[foundCount++] = elements[j];
      }
    }
    currentContext = found;
  }
  return currentContext;
}

/* That revolting regular expression explained 
/^(\w+)\[(\w+)([=~\|\^\$\*]?)=?"?([^\]"]*)"?\]$/
  \---/  \---/\-------------/    \-------/
    |      |         |               |
    |      |         |           The value
    |      |    ~,|,^,$,* or =
    |   Attribute 
   Tag
*/

//////////// End of presence check /////////////
}

src/main/resources/net/customware/confluence/plugin/linkvalidator/js/validator.js
//////////////// Start of presence check ///////////////
if(typeof LinkValidator=='undefined') {

var LinkValidator = {
	contextPath: "",
	
	validators: new Array(),
	currentValidator: 0,
	
	addValidator: function(validatorElement)
	{
		LinkValidator.validators[LinkValidator.validators.length] = validatorElement;
	},
	
	processNextValidator: function()
	{
		if (LinkValidator.currentValidator == LinkValidator.validators.length)
			return;
			
		var validator = LinkValidator.validators[LinkValidator.currentValidator++];
		
		var params = {
			url: validator.getAttribute("url"),
			verbose: validator.getAttribute("verbose"),
			timeout: validator.getAttribute("timeout")
		};
		
		LinkValidator.sendRequest("/plugins/linkvalidator/validate.action", params, validator);
	},
	
    sendRequest : function(uri, params, statusElement)
    {
        if (uri.charAt(0) == "/")
            uri = LinkValidator.contextPath + uri;

        var sparams = this.serializeParams(params);

        var callback =
        {
            method : "post",

            parameters : sparams,

            onComplete: function(data)
            {
                alert("sendRequest: response received");
                //debug("sendRequest: response received");
                if (data.status == 200)
                {
                    statusElement.innerHTML = data.responseText;
//                    Scaffold.View.handleResponse(data.responseXML);
                }
                else
                {
                    LinkValidator.showAjaxError(statusElement, data);
                }
                
                LinkValidator.processNextValidator();
            },

            onFailure : function(data, object)
            {
                //debug("sendRequest: failure");

                LinkValidator.showError(statusElement, "The operation failed.");

                LinkValidator.processNextValidator();
            },

            onException : function(data, exception)
            {
                //debug("sendRequest: exception");

                LinkValidator.showError(statusElement, "The operation failed: " + exception.getMessage());

                LinkValidator.processNextValidator();
            }
        }

        //debug("sendRequest: requesting '" + uri + "?" + sparams + "'");
        LinkValidator.showWorking(statusElement);

        // Send request to server
        //var req = new Ajax.Request(uri, callback);

        jQuery.ajax({
           url: uri,
           data: sparams
        }).complete(
            function(data){
                if (data.status == 200)
                {
                    statusElement.innerHTML = data.responseText;
                }
                else
                {
                    LinkValidator.showAjaxError(statusElement, data);
                }

                LinkValidator.processNextValidator();
            }
        );
        return false;
    },

	/**
	 * @param uri The main body of the URI.
	 * @param params The set of parameters to attach to the URI.
	 */
    serializeParams : function(params)
    {
        var key, value;
        var first = true;
        var uri = "";

        for (key in params)
        {
            var type = typeof(params[key]);
            if (params[key] != null && type != "function" && type != "undefined")
            {
                if (first)
                    first = false
                else
                    uri += "&";
                value = params[key];

                //debug("buildUri: key = '" + key + "'; value = '" + value + "'; type = " + typeof(params[key]));

                uri += encodeURIComponent(key) + "=" + encodeURIComponent(value);
            }
        }

        return uri;
    },
    

    showAjaxError : function(statusElement, data)
    {
        var message = "An error occurred while processing. Please contact an administrator.";
        switch (data.status)
            {
            case 500:
                message = "There was an error on the server. Please check your server logs for details.";
                break;
            case 404:
                message = "The URL for the requested action does not exist.";
                break;
        }
        LinkValidator.showError(statusElement, message);
    },
    
    showError : function(statusElement, message)
    {
    	LinkValidator.showIcon(statusElement, LinkValidator.getIconPath("emoticons/error.gif"), message);
    },
    
    showWorking : function(statusElement)
    {
    	LinkValidator.showIcon(statusElement, LinkValidator.getIconPath("wait.gif"), "Working...");
    },

    showIcon : function(statusElement, src, message)
    {
        var icon = "<img src='" + src + "' text='" + message + "'/>";
        if (statusElement.getAttribute("verbose") == "true")
        	icon += " " + message;
        statusElement.innerHTML = icon;
    },
    
    getIconPath : function(iconName)
    {
    	return LinkValidator.contextPath + "/images/icons/" + iconName;
    },
    
    rules : {
        '.validatorElement' : function(el)
        {
			LinkValidator.addValidator(el);
        }
    }
};

Behaviour.register(LinkValidator.rules);

Behaviour.addLoadEvent(LinkValidator.processNextValidator);
if (navigator.appName == "Microsoft Internet Explorer") {
Behaviour.start();
}
//////////// End of presence check /////////////
}

src/main/java/net/customware/confluence/plugin/linkvalidator/LinkValidatorMacro.java
package net.customware.confluence.plugin.linkvalidator;

import com.atlassian.confluence.content.render.xhtml.ConversionContext;
import com.atlassian.confluence.content.render.xhtml.DefaultConversionContext;
import com.atlassian.confluence.core.ContextPathHolder;
import com.atlassian.confluence.macro.Macro;
import com.atlassian.confluence.macro.MacroExecutionException;
import com.atlassian.confluence.renderer.radeox.macros.MacroUtils;
import com.atlassian.confluence.util.velocity.VelocityUtils;
import com.atlassian.renderer.RenderContext;
import com.atlassian.renderer.v2.RenderMode;
import com.atlassian.renderer.v2.macro.BaseMacro;
import com.atlassian.renderer.v2.macro.MacroException;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.Map;

import static com.atlassian.confluence.util.GeneralUtil.escapeXml;
import static org.apache.commons.lang3.StringUtils.isBlank;
import static org.apache.commons.lang3.StringUtils.isNotBlank;

public class LinkValidatorMacro extends BaseMacro implements Macro {

    private ContextPathHolder contextPathHolder;

    @Override
    public String execute(Map map, String s, RenderContext renderContext) throws MacroException {
        try {
            return execute(map, s, new DefaultConversionContext(renderContext));
        } catch (MacroExecutionException e) {
            throw new MacroException(e);
        }
    }

    @Override
    public String execute(Map<String, String> params, String body, ConversionContext conversionContext) throws MacroExecutionException {
        String link = isNotBlank(params.get("url")) ? params.get("url") : null;
        if (isBlank(link)) {
            throw new MacroExecutionException("Please supply the URL to validate.");
        }

        // If true, output the message in full.
        boolean verbose = isNotBlank(params.get("verbose")) ? Boolean.valueOf(params.get("verbose")) : false;
        // The timeout in seconds.
        int timeout = isNotBlank(params.get("timeout")) ? Integer.valueOf(params.get("timeout")) : 5;

        Map<String,Object> contextMap = MacroUtils.defaultVelocityContext();
        contextMap.put("contextPath", contextPathHolder.getContextPath());
        contextMap.put("url", escapeXml(link));
        contextMap.put("timeout", timeout);
        contextMap.put("verbose", verbose);
        contextMap.put("icon", LinkValidatorUtil.createUnknownIcon("Unvalidated", verbose));

        return VelocityUtils.getRenderedTemplate("net/customware/confluence/plugin/linkvalidator/macro/link-validator.vm", contextMap);
    }

    public BodyType getBodyType() {
        return BodyType.NONE;
    }

    public OutputType getOutputType() {
        return OutputType.INLINE;
    }

    @Autowired
    public void setContextPathHolder(ContextPathHolder contextPathHolder) {
        this.contextPathHolder = contextPathHolder;
    }

    @Override
    public boolean hasBody() {
        return true;
    }

    @Override
    public RenderMode getBodyRenderMode() {
        return RenderMode.suppress(RenderMode.F_FIRST_PARA + RenderMode.F_LINKS);
    }
}


src/main/java/net/customware/confluence/plugin/linkvalidator/LinkValidatorMacroMigration.java
package net.customware.confluence.plugin.linkvalidator;

import com.atlassian.confluence.content.render.xhtml.ConversionContext;
import com.atlassian.confluence.macro.xhtml.MacroMigration;
import com.atlassian.confluence.xhtml.api.MacroDefinition;
import org.apache.commons.lang.StringUtils;

import java.util.HashMap;
import java.util.Map;

public class LinkValidatorMacroMigration implements MacroMigration {

    public MacroDefinition migrate(MacroDefinition macroDefinition, ConversionContext conversionContext) {
        Map<String, String> params = macroDefinition.getParameters();
        String url = params.get("url");
        if (StringUtils.isNotEmpty(url)) {
            return macroDefinition;
        }
        url = macroDefinition.getDefaultParameterValue();
        if (StringUtils.isEmpty(url)) {
            url = macroDefinition.getBodyText();
        }

        Map<String, String> newParams = new HashMap<String, String>();
        newParams.putAll(params);
        if (url != null) {
            newParams.put("url", url.trim());
        }
        return new MacroDefinition(macroDefinition.getName(), null, null, newParams);
    }

    /**
     {link-validator:verbose=true|timeout=10}
     http://www.google.com
     {link-validator}
     {link-validator:url=http://www.google.com|verbose=true|timeout=10}
     {link-validator:http://www.google.com|verbose=true|timeout=10}
     */
}


src/main/java/net/customware/confluence/plugin/linkvalidator/LinkValidatorUtil.java
package net.customware.confluence.plugin.linkvalidator;

import com.atlassian.confluence.setup.BootstrapManager;
import com.atlassian.confluence.util.GeneralUtil;
import com.atlassian.spring.container.ContainerManager;
import org.apache.http.client.config.RequestConfig;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.servlet.http.HttpServletRequest;
import java.io.IOException;
import java.net.URISyntaxException;
import java.net.URL;

public class LinkValidatorUtil
{
    private static final Logger LOG                         = LoggerFactory.getLogger(LinkValidatorUtil.class);

    private static final String HTTPS_PROTOCOL              = "https";

    private static final String HTTP_PROTOCOL               = "http";

    private static final String INFO_ICON                   = "emoticons/information.gif";

    private static final String GOOD_ICON                   = "emoticons/check.gif";

    private static final String ERROR_ICON                  = "emoticons/warning.gif";

    private static final String REDIRECT_ICON               = "ref_16.gif";

    private static final String UNKNOWN_ICON                = "emoticons/help_16.gif";

    private static final String FORBIDDEN_ICON              = "emoticons/forbidden.gif";

    private static final String LINK_VALIDATOR_HEADER_VALUE = "1";

    private static final String LINK_VALIDATOR_HEADER       = "x-randombits-link-validator";

    static BootstrapManager     bootstrapManager;

    private LinkValidatorUtil()
    {
    }

    public static boolean isBeingValidated(HttpServletRequest request)
    {
        return LINK_VALIDATOR_HEADER_VALUE.equals(request.getHeader(LINK_VALIDATOR_HEADER));
    }

    public static String validate(URL url, boolean verbose, int timeout)
    {
        CloseableHttpClient httpClient = HttpClients.createDefault();
        RequestConfig requestConfig  = RequestConfig.custom()
                .setSocketTimeout(timeout * 1000)
                .build();
        CloseableHttpResponse response = null;

        try {
            // Check that we're only validating HTTP(S) locations.
            if (!HTTP_PROTOCOL.equals(url.getProtocol()) && !HTTPS_PROTOCOL.equals(url.getProtocol())) {
                return createErrorIcon("Unsupported protocol: " + url.getProtocol(), verbose);
            }

            LOG.debug("connecting to url: " + url);

            HttpGet httpGet = new HttpGet(url.toURI());
            httpGet.setConfig(requestConfig);
            response = httpClient.execute(httpGet);

            return createHttpStatusIcon(response.getStatusLine().getStatusCode(), response.getStatusLine().getReasonPhrase(), verbose);
        } catch (URISyntaxException e) {
            LOG.error("Invalid URL: " + url.toString(), e);
            return createErrorIcon("Invalid URL: " + e, verbose);
        } catch (IOException e) {
            LOG.error("Failed to execute HTTP GET to URL: " + url.toString(), e);
            return createErrorIcon(e.getLocalizedMessage(), verbose);
        } finally {
            if (response != null) {
                try {
                    response.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    static String createHttpStatusIcon(int responseCode, String responseMessage, boolean verbose)
    {
        String icon = UNKNOWN_ICON;
        if (responseCode >= 100 && responseCode < 200)
            icon = INFO_ICON;
        else if (responseCode < 300)
            icon = GOOD_ICON;
        else if (responseCode < 400)
            icon = REDIRECT_ICON;
        else if (responseCode < 500)
            icon = FORBIDDEN_ICON;
        else if (responseCode < 600) icon = ERROR_ICON;

        return createIcon(icon, responseMessage + " (" + responseCode + ")", verbose);
    }

    private static String createIcon(String iconPath, String status, boolean verbose)
    {
        StringBuffer out = new StringBuffer();

        // FIXME [20130402 YKT] Please remove the following method, if you have found a better way to resolve icon path
        // Could not find any confluence API that able the resolve icon path that backward compatible.
        iconPath = resolvedIconExtensions(iconPath);

        out.append("<img");
        out.append(" src='").append(getContextPath()).append("/images/icons/").append(iconPath).append("'");
        if (status != null) out.append(" title='").append(GeneralUtil.escapeXml(status)).append("'");
        out.append("/>");

        if (verbose && status != null) out.append(" ").append(GeneralUtil.escapeXml(status));

        return out.toString();
    }

    private static String getContextPath()
    {
        return getBootstrapManager().getWebAppContextPath();
    }

    public static BootstrapManager getBootstrapManager()
    {
        if (bootstrapManager == null)
            bootstrapManager = (BootstrapManager) ContainerManager.getComponent("bootstrapManager");
        return bootstrapManager;
    }

    public static String createInfoIcon(String status, boolean verbose)
    {
        return createIcon(INFO_ICON, status, verbose);
    }

    public static String createGoodIcon(String status, boolean verbose)
    {
        return createIcon(GOOD_ICON, status, verbose);
    }

    public static String createUnknownIcon(String status, boolean verbose)
    {
        return createIcon(UNKNOWN_ICON, status, verbose);
    }

    public static String createErrorIcon(String status, boolean verbose)
    {
        return createIcon(ERROR_ICON, status, verbose);
    }

    public static String createRedirectIcon(String status, boolean verbose)
    {
        return createIcon(REDIRECT_ICON, status, verbose);
    }

    public static String createForbiddenIcon(String status, boolean verbose)
    {
        return createIcon(FORBIDDEN_ICON, status, verbose);
    }

    /**
     * Return a given icon path to a converted one (to {@code .png}) from any other formats.
     * It is a behavior introduced since Confluence 5 that all icons are created as PNG.
     */
    private static String resolvedIconExtensions(String iconPath){

        if(GeneralUtil.getVersionNumber().compareToIgnoreCase("5.0") >= 0){
            int lastDotPos = iconPath.lastIndexOf('.');
            if( lastDotPos != -1){
                return iconPath.substring(0, lastDotPos) + ".png";
            }
        }
        return iconPath;
    }

}
